class A over Exception {
    pub self () {}
}

class B over Exception {
    pub self () {}
}

class C over B {
    pub self () {}
}

fn foo (i : i32)-> i32
    throws A, B, C;

fn onA ();
fn onB ();
fn onC ();

fn main ()
{
    let _ = {
        foo (0)
    } catch {
        B ()  => {
            onB ();
            4
        }
        C () => {
            onC ();
            5
        }
        A () => {
            onA ();
            3
        }
    };
}
